File: architecture/serialization-formats/java/src/main/java/tech/khorost/serialization/Exchange.java
package tech.khorost.serialization;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.HexFormat;

/**
 * Обмен байтами между двумя реализациями пробы через файлы рабочего
 * каталога прогона (Задача 8: ось перекрёстного чтения — schemas/spec.md,
 * раздел «Ось 4»).
 *
 * Одна реализация не может декодировать байты, которые закодировала
 * другая, внутри одного процесса: перекрёстная проверка требует, чтобы
 * одна реализация ЗАПИСАЛА байты на диск, а другая — независимо от
 * первой — их ПРОЧИТАЛА. Каталог обмена не аргумент пробы (§4.2 не
 * нарушается): его выбирает вызывающий сценарий (в проде — текущий
 * рабочий каталог процесса), а имя файла внутри него полностью выводится
 * из координат клетки, номера записи и языка-писателя.
 *
 * Три ловушки, которые обязан ловить этот класс (решение контроллера
 * Задачи 8, то же самое решение, что и в Go-части —
 * internal/exchange/exchange.go, — оба обязаны совпадать по имени файла
 * и форме конверта, иначе писатель и читатель не найдут друг друга):
 * перепутанное имя (координаты внутри файла не совпадают с запрошенными),
 * недописанный или испорченный файл (дайджест не сходится с содержимым),
 * и след прошлого прогона — это уже забота вызывающего сценария, каталог
 * обмена обязан быть очищен ДО прогона, класс сам не отличит стухший файл
 * от свежего с теми же координатами.
 */
final class Exchange {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private Exchange() {}

    /**
     * Детерминированное имя файла обмена: функция только от координат
     * клетки, номера записи и языка-писателя. Обязано давать то же самое
     * имя, что и Go-часть, на тех же аргументах.
     */
    static Path crossFileName(Path dir, String lang, String format, String change, String direction, int recordIndex) {
        String name = String.format("cross_%s_%s_%s_%s_%d.json", lang, format, change, direction, recordIndex);
        return dir.resolve(name);
    }

    /**
     * Пишет конверт АТОМАРНО: сначала во временный файл того же
     * каталога, затем переименовывает. Обрыв процесса посередине записи
     * не оставит недописанный файл ПОД ОЖИДАЕМЫМ ИМЕНЕМ — readCross либо
     * увидит целый файл, либо не увидит никакого.
     */
    static void writeCross(Path dir, String lang, String format, String change, String direction,
                            int recordIndex, byte[] bytes) {
        ObjectNode env = MAPPER.createObjectNode();
        env.put("lang", lang);
        env.put("format", format);
        env.put("change", change);
        env.put("direction", direction);
        env.put("record_index", recordIndex);
        env.put("sha256", sha256Hex(bytes));
        env.put("bytes_b64", Base64.getEncoder().encodeToString(bytes));

        Path target = crossFileName(dir, lang, format, change, direction, recordIndex);
        Path tmp = null;
        try {
            Files.createDirectories(dir);
            tmp = Files.createTempFile(dir, "cross-", ".tmp");
            Files.write(tmp, MAPPER.writeValueAsBytes(env));
            Files.move(tmp, target, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
        } catch (IOException e) {
            if (tmp != null) {
                try {
                    Files.deleteIfExists(tmp);
                } catch (IOException ignore) {
                    // Лучшее из возможного — сам временный файл уже не
                    // тот артефакт, который проверяет остальной класс.
                }
            }
            throw new Failures.ProbeRefusal("exchange: запись файла обмена " + target + ": " + e);
        }
    }

    /**
     * Читает файл обмена по детерминированному имени и возвращает байты
     * только после того, как:
     * <ol>
     * <li>файл найден и разобран как конверт;</li>
     * <li>координаты ВНУТРИ конверта совпадают с запрошенными (лечит
     *     перепутанное имя);</li>
     * <li>дайджест содержимого совпадает с зафиксированным при записи
     *     (лечит недописанный или испорченный файл).</li>
     * </ol>
     * Файл обмена — единственная связь между двумя независимыми
     * процессами, и она не защищена ничем, кроме этой проверки.
     */
    static byte[] readCross(Path dir, String lang, String format, String change, String direction, int recordIndex) {
        Path path = crossFileName(dir, lang, format, change, direction, recordIndex);
        byte[] raw;
        try {
            raw = Files.readAllBytes(path);
        } catch (IOException e) {
            throw new Failures.ProbeRefusal("exchange: файл обмена " + path + " не прочитан: " + e);
        }
        JsonNode env;
        try {
            env = MAPPER.readTree(raw);
        } catch (IOException e) {
            throw new Failures.ProbeRefusal("exchange: файл обмена " + path
                    + " повреждён: не разбирается как JSON: " + e);
        }
        String gotLang = text(env, "lang");
        String gotFormat = text(env, "format");
        String gotChange = text(env, "change");
        String gotDirection = text(env, "direction");
        int gotIndex = env.has("record_index") ? env.get("record_index").asInt() : Integer.MIN_VALUE;
        if (!lang.equals(gotLang) || !format.equals(gotFormat) || !change.equals(gotChange)
                || !direction.equals(gotDirection) || recordIndex != gotIndex) {
            throw new Failures.ProbeRefusal("exchange: координаты внутри файла обмена " + path
                    + " не совпадают с запрошенными (лежит: lang=" + gotLang + " format=" + gotFormat
                    + " change=" + gotChange + " direction=" + gotDirection + " record_index=" + gotIndex
                    + "; ожидались: lang=" + lang + " format=" + format + " change=" + change
                    + " direction=" + direction + " record_index=" + recordIndex
                    + ") — файл подменён или взят не тот");
        }
        byte[] bytes;
        try {
            bytes = Base64.getDecoder().decode(text(env, "bytes_b64"));
        } catch (IllegalArgumentException | NullPointerException e) {
            throw new Failures.ProbeRefusal("exchange: файл обмена " + path
                    + " повреждён: base64 не разбирается: " + e);
        }
        String expected = text(env, "sha256");
        String actual = sha256Hex(bytes);
        if (!actual.equals(expected)) {
            throw new Failures.ProbeRefusal("exchange: файл обмена " + path
                    + " повреждён или недописан: дайджест не совпадает (записан " + expected
                    + ", посчитан " + actual + ")");
        }
        return bytes;
    }

    private static String text(JsonNode node, String field) {
        JsonNode v = node.get(field);
        return v == null ? null : v.asText();
    }

    private static String sha256Hex(byte[] data) {
        try {
            return HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(data));
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: architecture/serialization-formats/go/internal/exchange/exchange.go
// Package exchange реализует передачу байтов между двумя реализациями
// пробы через файлы рабочего каталога прогона (Задача 8: ось
// перекрёстного чтения — schemas/spec.md, раздел «Ось 4»).
//
// Одна реализация не может декодировать байты, которые закодировала
// другая, внутри одного процесса: перекрёстная проверка требует, чтобы
// одна реализация ЗАПИСАЛА байты на диск, а другая — независимо от
// первой — их ПРОЧИТАЛА. Каталог обмена — не аргумент пробы (§4.2 не
// нарушается): это ТЕКУЩИЙ РАБОЧИЙ КАТАЛОГ ПРОЦЕССА, который выбирает
// вызывающий сценарий, а имя файла внутри него полностью выводится из
// координат клетки, номера записи и языка-писателя — не из содержимого
// и не из произвольного счётчика.
//
// Три ловушки, которые обязан ловить этот пакет (решение контроллера
// Задачи 8): перепутанное имя (координаты внутри файла не совпадают с
// запрошенными), недописанный или испорченный файл (дайджест не сходится
// с содержимым) и запись, оставшаяся от прошлого прогона (это уже забота
// вызывающего сценария — каталог обмена обязан быть очищен ДО прогона,
// см. bench/run-cross.sh; сам пакет очисткой не занимается, потому что
// не может отличить «стухший» файл от свежего с теми же координатами).
package exchange

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// crossEnvelope — самоописывающийся конверт с байтами одной канонической
// записи. Поля Format/Change/Direction/RecordIndex/Lang дублируют то, что
// уже зашито в имени файла, — не для экономии, а потому что имя файла
// проверяет только то, ПО КАКОМУ ПУТИ прочитан файл, а конверт проверяет,
// ЧТО в нём реально лежит. Расхождение между этими двумя источниками и
// есть «перепутанное имя».
type crossEnvelope struct {
	Lang        string `json:"lang"` // язык-писатель
	Format      string `json:"format"`
	Change      string `json:"change"`
	Direction   string `json:"direction"`
	RecordIndex int    `json:"record_index"`
	SHA256      string `json:"sha256"`    // дайджест байтов ДО base64
	BytesB64    string `json:"bytes_b64"` // сами байты, закодированные в base64 ради текстового контейнера
}

// CrossFileName — детерминированное имя файла обмена: функция только от
// координат клетки, номера записи и языка-писателя. Один и тот же вызов
// с одними и теми же аргументами в любой реализации обязан дать одно и
// то же имя — иначе писатель и читатель не найдут друг друга.
func CrossFileName(dir, lang, format, change, direction string, recordIndex int) string {
	name := fmt.Sprintf("cross_%s_%s_%s_%s_%d.json", lang, format, change, direction, recordIndex)
	return filepath.Join(dir, name)
}

// WriteCross кодирует байты в конверт и пишет его АТОМАРНО: сначала во
// временный файл в том же каталоге, затем переименовывает. Не пишущий
// напрямую в целевое имя, обрыв процесса посередине записи не оставит
// недописанный файл ПОД ОЖИДАЕМЫМ ИМЕНЕМ — ReadCross либо увидит целый
// файл, либо не увидит никакого.
func WriteCross(dir, lang, format, change, direction string, recordIndex int, b []byte) error {
	sum := sha256.Sum256(b)
	env := crossEnvelope{
		Lang:        lang,
		Format:      format,
		Change:      change,
		Direction:   direction,
		RecordIndex: recordIndex,
		SHA256:      hex.EncodeToString(sum[:]),
		BytesB64:    base64.StdEncoding.EncodeToString(b),
	}
	raw, err := json.Marshal(env)
	if err != nil {
		return fmt.Errorf("exchange: конверт не сериализуется: %w", err)
	}

	target := CrossFileName(dir, lang, format, change, direction, recordIndex)
	tmp, err := os.CreateTemp(dir, "cross-*.tmp")
	if err != nil {
		return fmt.Errorf("exchange: временный файл обмена: %w", err)
	}
	tmpName := tmp.Name()
	if _, err := tmp.Write(raw); err != nil {
		tmp.Close()
		os.Remove(tmpName)
		return fmt.Errorf("exchange: запись временного файла обмена: %w", err)
	}
	if err := tmp.Close(); err != nil {
		os.Remove(tmpName)
		return fmt.Errorf("exchange: закрытие временного файла обмена: %w", err)
	}
	if err := os.Rename(tmpName, target); err != nil {
		os.Remove(tmpName)
		return fmt.Errorf("exchange: переименование файла обмена: %w", err)
	}
	return nil
}

// ReadCross читает файл обмена по детерминированному имени, выведенному
// ИЗ ТЕХ ЖЕ координат, и возвращает байты только после того, как:
//  1. файл найден и разобран как конверт;
//  2. координаты ВНУТРИ конверта совпадают с запрошенными (лечит
//     перепутанное имя: содержимое не то, что ожидает читатель, даже
//     если оно лежит по правильному пути);
//  3. дайджест содержимого совпадает с зафиксированным при записи (лечит
//     недописанный или испорченный файл).
//
// Ни один шаг не пропускается ради скорости: файл обмена — единственная
// связь между двумя независимыми процессами, и она не защищена ничем,
// кроме этой проверки.
func ReadCross(dir, lang, format, change, direction string, recordIndex int) ([]byte, error) {
	path := CrossFileName(dir, lang, format, change, direction, recordIndex)
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("exchange: файл обмена %s не прочитан: %w", path, err)
	}
	var env crossEnvelope
	if err := json.Unmarshal(raw, &env); err != nil {
		return nil, fmt.Errorf("exchange: файл обмена %s повреждён: не разбирается как JSON: %w", path, err)
	}
	if env.Lang != lang || env.Format != format || env.Change != change ||
		env.Direction != direction || env.RecordIndex != recordIndex {
		return nil, fmt.Errorf(
			"exchange: координаты внутри файла обмена %s не совпадают с запрошенными "+
				"(лежит: lang=%s format=%s change=%s direction=%s record_index=%d; "+
				"ожидались: lang=%s format=%s change=%s direction=%s record_index=%d) — "+
				"файл подменён или взят не тот",
			path, env.Lang, env.Format, env.Change, env.Direction, env.RecordIndex,
			lang, format, change, direction, recordIndex)
	}
	b, err := base64.StdEncoding.DecodeString(env.BytesB64)
	if err != nil {
		return nil, fmt.Errorf("exchange: файл обмена %s повреждён: base64 не разбирается: %w", path, err)
	}
	sum := sha256.Sum256(b)
	got := hex.EncodeToString(sum[:])
	if got != env.SHA256 {
		return nil, fmt.Errorf(
			"exchange: файл обмена %s повреждён или недописан: дайджест не совпадает "+
				"(записан %s, посчитан %s)", path, env.SHA256, got)
	}
	return b, nil
}
